{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f6c82f",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "## Assignment 4\n",
    "\n",
    "A wig format is used for genome coverage of a dataset. The declaration line starts with the word variableStep and is followed by a specification for a chromosome. Ignore the optional span parameter for now. After the declaration line for each chromosome, each line has two columns. First column is chromosome position, and second is the coverage at that position. You have to write the code to read the \"80mM\\_Ctl\\_sub.wig\" file and then print out the coverage as three separate values:\n",
    "\n",
    "Chromosome Position Value\n",
    "\n",
    "Here are the steps to code:\n",
    "\n",
    "1. First click on the wig file to open it in text editor and inspect the format\n",
    "2. Write code to read the wig file line by line\n",
    "3. When you see the declaration line, extract the string representing chromosome \\(\"2L\"\\) and store it in a variable\n",
    "4. All other lines, split into two \"words\" \\- one for position and the other for coverage.\n",
    "5. For each line, print chromosome, position, and coverage\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "id": "5a4153",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
    "# write code here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d3a7f",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "For the fastq file that we worked on, calculate the GC content at each position of the read. Here is a demo:\n",
    "\n",
    "`sequence 1: ATGCATGC` \n",
    "\n",
    "`sequence 2: GAGAGCTA`\n",
    "\n",
    "First position is 50% GC, second position is 0% GC and so on.\n",
    "\n",
    "Here are the steps to code:\n",
    "\n",
    "1. Declare a list to store the count of G/C nucleotides at each position in the sequence \\(the indices of the string\\).\n",
    "2. With each sequence line in the fastq file\n",
    "   1. For each position in the sequence string, add 1 to the list at that position if the position in the string is G or C\n",
    "   2. Keep count of the number of sequence lines\n",
    "3. Divide each element of the list by the total number of sequences and print the value\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "id": "9cab25",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
    "# write code here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "189f91",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Read a uniprot text file \"HistoneH3_Uniprot.txt\". Write code to extract just the subcellular localization and the amino acid sequence from the uniprot file.\n",
    "\n",
    "Here are the steps to code:\n",
    "\n",
    "1. First click on HistoneH3\\_Uniprot.txt to open it in a text editor and examine it visually \\- what are the flags that identify subcellular location and the amino acid sequence?\n",
    "   1. Since amino acid sequence is in multiple lines, you can use a \"sentinel variable\" to alert the code if you are in the right part of the file.\n",
    "2. In your code, set the sentinal variable to 0 \\(let us say f=0\\)\n",
    "3. Write code to read the file line by line and use conditional statements to check for presence of flags you identified in step 1\n",
    "4. If the flag for subcellular location is identified, extract the necessary string for subcellular location\n",
    "5. If flag is identified that says the next lines are going to be the amino acid sequence, set the sentinel variable to an arbitrary value \\(for example, set f=1\\)\n",
    "   1. If f==1, then print that line\n",
    "6. If the flag that says sequence lines are done is identified, set the sentinel variable back to 0 \\(f=0\\) \\- your code will stop printing sequence lines.\n",
    "7. The key is to play around with the order of setting the sentinel variable and printing lines so that you print only the sequence lines and not the lines before or after that specify start and end of sequence lines.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "id": "c6bfb3",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
    "# write code here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "feeb89",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "In the fasta file \"protein\\_sequences.fasta\", calculate the net charge of the protein. The negatively charged residues are 'D' and  'E'. The output should look like \\(These are made up numbers\\):\n",
    "\n",
    "Sequence_name:HuR Net_charge:4 \n",
    "\n",
    "Here are the steps to code:\n",
    "\n",
    "1. Start with the code we discussed in class and extend it to count negatively charged residues in a separate variable\n",
    "2. Subtract the total number of negatively charged variables from the total number of positively charged variables \n",
    "3. Some things to consider: \n",
    "   1. The sequence for each protein is in multiple lines \\- you need to account for all lines for each protein when calculating the net charge. \n",
    "   2. Make sure your code calculates the charge for all 4 proteins\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "id": "d3771c",
   "metadata": {
    "collapsed": false
   },
   "outputs": [
   ],
   "source": [
    "# write code here"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {
  },
  "kernelspec": {
   "argv": [
    "/ext/venvs/cocalc/bin/python3",
    "-Xfrozen_modules=off",
    "-m",
    "ipykernel_launcher",
    "--HistoryManager.enabled=False",
    "--matplotlib=inline",
    "-c",
    "%config InlineBackend.figure_formats = set(['retina'])\nimport matplotlib; matplotlib.rcParams['figure.figsize'] = (12, 7)",
    "-f",
    "{connection_file}"
   ],
   "display_name": "Python 3 (CoCalc)",
   "env": {
   },
   "language": "python",
   "metadata": {
    "cocalc": {
     "description": "Python 3 system-wide environment. Use 'CoCalc' for the most complete Python environment.",
     "priority": 100,
     "url": "https://www.python.org/"
    },
    "debugger": true
   },
   "name": "python3",
   "resource_dir": "/ext/jupyter/kernels/python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}